skyd/skyd.go

```diff
@@ -1,16 +1,137 @@
 package skyd
 
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/node/api/client"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+var (
+	// skylinksCache is a local cache of the list of skylinks pinned by the
+	// local skyd
+	skylinksCache pinnedSkylinksCache
+	// skylinksCacheDuration defines the duration of skylinksCache
+	skylinksCacheDuration = time.Hour
+)
+
 type (
 	// Client allows us to call the local skyd instance.
 	Client struct {
+		staticClient *client.Client
+	}
+	// ClientInterface describes the interface exposed by Client, so we can
+	// mock it for testing.
+	ClientInterface interface {
+		Pin(skylink string) error
+		PinnedSkylinks() (skylinks []string, err error)
+		Unpin(skylink string) error
+	}
+
+	// renterDirCache is a simple cache of the renter's directory information,
+	// so we don't need to fetch that for each skylink we potentially want to
+	// pin/unpin.
+	pinnedSkylinksCache struct {
+		Skylinks   []string
+		Expiration time.Time
+		mu         sync.Mutex
 	}
 )
 
-/*
-TODO
- - env vars for talking to skyd - host, port, apipassword, etc.
- - methods for calling skyd:
-	- list all pinned skylinks
-	- pin a skylink
-	- unpin a skylink
-*/
+// NewClient creates a new skyd client.
+func NewClient(host, port, password string) *Client {
+	opts := client.Options{
+		Address:       fmt.Sprintf("%s:%s", host, port),
+		Password:      password,
+		UserAgent:     "Sia-Agent",
+		CheckRedirect: nil,
+	}
+	return &Client{
+		staticClient: client.New(opts),
+	}
+}
+
+// Pin instructs the local skyd to pin the given skylink.
+func (c *Client) Pin(skylink string) error {
+	pinned, err := c.isPinned(skylink)
+	if err != nil {
+		return err
+	}
+	if pinned {
+		// The skylink is already locally pinned, nothing to do.
+		return nil
+	}
+	sp, err := skymodules.NewSiaPath(skylink)
+	if err != nil {
+		return errors.AddContext(err, "failed to create siapath")
+	}
+	spp := skymodules.SkyfilePinParameters{
+		SiaPath: sp,
+	}
+	return c.staticClient.SkynetSkylinkPinPost(skylink, spp)
+}
+
+// PinnedSkylinks returns the list of skylinks pinned by the local skyd.
+func (c *Client) PinnedSkylinks() (skylinks []string, err error) {
+	skylinksCache.mu.Lock()
```

Comment:
I think that in this case it's fine but I'm ready to be convinced in the opposite.

The reason I think it's not only OK but also helpful is that we only do I/O while holding a lock if the cache is expired and needs to be refreshed. If we have another thread that calls this method while the cache is being refreshed, we have two cases:
1. If we're holding a lock on the cache, the second thread will block and wait. Once it unblocks, it will check the expiration of the cache and it's going to find that it's valid.
2. If we're not holding a lock (we release it after we find that the cache is expired), then second thread is going to make the same check as the first thread. It's going to find that the cache is expired, it's going to release the lock and fetch the data from `skyd`. In the end it's going to lock the cache and overwrite it, even though the cache would have already been updated by the first thread. This will essentially cause a lot of unnecessary work.

Reply:
so for example the `Pin` and the `Unpin` methods both update the cache status, which both try and acquire the cache lock, which might be in the middle of rebuilding an expired cache, which might take minutes, and I think it might take a really long time

I just think there's a reasonable possibility the service might be faulty for 5-10 minutes every hour because we're holding this lock while doing a bunch of I/O.

Reply:
Let's take a step back for a moment. What we're trying to do here is prevent data duplication. So we can ignore `Unpin` where I've already removed the check of the cache - we can safely call that every time.

How can we get data duplication? We need the skylink (1) to be uploaded by another server and (2) to not be marked as uploaded by the current one. 
(2) can happen if the current server hasn't been swept yet, so the DB doesn't know what it's pinning. This is easily fixable by populating the cache once and then updating it each time Pinned gets notified that the current server is pinning  a new skylink. The problem here is that if something unpins a skylink from skyd, Pinner won't know about it and when scanning for underpinned skylinks it will still think that the current server is pinning it, potentially leaving a skylink underpinned until it dies. Our mechanism for countering that are the regular sweeps that will synchronise the DB with the actual list of skylinks pinned by skyd.

This makes me think that we can get away with rebuilding the cache just once before each scan and assuming that it's acceptably up-to-date for the duration of the scan.

Another thing to keep in mind is that we want to make this listing of files pinned by skyd available via skyd's API, which should make it relatively fast. It will still mean transferring hundreds of MBs of data but it won't be as expensive as the current approach. I am not sure this will affect our approach but it's good to keep in mind, anyway.

Reply:
One solution is to do what we do with other caches which is the have the update happening in a background thread. So pin and unpin are always just pulling from the in memory cache.

if the cache is out of date, it can pull the new data without holding a lock. Then once it has all the new data, it acquires the lock and updates the cache. 

that would reduce the lock conflict times.

I would recommend handling this in a follow up to allow better review of this specific process.

Reply:
A follow up sounds good here. I agree that the current approach is not the best possible. I'm not even sure about using a cache at all because it might lead to data duplication. What would make me happiest would be for `skyd` to keep tabs on what it's pinning, so I can ask it each time before pinning something. This would need to be efficient, though, and it currently doesn't exist.

@peterjan are you fine with a F/U on this or do you feel like it's an immediate issue?

Reply:
Yeah sure, I don't mind. I think a background process updating in the background is easiest. It's also quick to implement.